controller/src/main/java/io/pravega/controller/store/stream/ZKStreamMetadataStore.java

```diff
@@ -86,15 +113,88 @@
+                                return list.subList(0, list.size() - 2);
+                            } else {
+                                return new ArrayList<Long>();
+                            }
+                        }
+                )
+                // region Backward Compatibility.
+                // TODO 2755 retire code in this region 
+                .thenCompose(toDeleteList -> {
+                    CompletableFuture<Void> future;
+                    if (!Config.DISABLE_COMPLETED_TXN_BACKWARD_COMPATIBILITY) {
+                        log.debug("deleting batches {} on old scheme" + toDeleteList);
+
+                        // For backward compatibility, find full path on "toDelete" and find corresponding paths in old 
+                        // scheme and delete each txn individually!!
+                        future = Futures.allOf(
+                                toDeleteList.stream()
+                                            .map(toDelete -> deleteOldSchemeTxns(String.format(COMPLETED_TX_BATCH_PATH, toDelete)))
+                                            .collect(Collectors.toList()));
+                    } else {
+                        future = CompletableFuture.completedFuture(null);
+                    }
+                    return future.thenApply(x -> toDeleteList);
+
+                })
+                // endregion
+                .thenCompose(toDeleteList -> {
+                    log.debug("deleting batches {} on new scheme" + toDeleteList);
+
+                    // delete all those marked for toDelete.
+                    return Futures.allOf(toDeleteList.stream()
+                            .map(toDelete -> storeHelper.deleteTree(String.format(COMPLETED_TX_BATCH_PATH, toDelete)))
+                            .collect(Collectors.toList()));
+                });
+    }
+
+    // region Backward Compatibility.
+    // TODO 2755 retire code in this region 
+    private CompletableFuture<List<Void>> deleteOldSchemeTxns(String rootPath) {
+        // getAllChildren will return all scopes
+        return storeHelper.getChildren(rootPath)
+                .thenCompose(scopes -> {
+                    return Futures.allOfWithResults(scopes.stream().map(scope -> {
+                        String scopePath = ZKPaths.makePath(rootPath, scope);
+                        return storeHelper.getChildren(scopePath)
+                                .thenCompose(streams -> Futures.allOfWithResults(streams.stream().map(stream -> {
+                                    String streamPath = ZKPaths.makePath(scopePath, stream);
+                                    return storeHelper.getChildren(streamPath)
+                                            .thenApply(txnIds -> txnIds.stream().map(txnId -> {
+                                                String txnPath = getOldSchemePath(scope, stream, txnId);
+                                                return storeHelper.deletePath(txnPath, false);
```

Comment:
This really doesn't look like it will scale.

Reply:
we can remove clean up old scheme completely from the code(this inefficient clean up)..and give admins a tool/rest api for cleaning up txns on old schema whenever they know they have upgraded all controllers and data in old scheme is stale.

Reply:
If in the mean time it OOMs the process we still have an issue.